akka-remote/src/main/scala/akka/remote/Remoting.scala

```diff
@@ -548,18 +555,28 @@ private[remote] class EndpointManager(conf: Config, log: LoggingAdapter) extends
             "address cannot be quarantined without knowing the UID, gating instead for {} ms.",
             address, settings.RetryGateClosedFor.toMillis)
           endpoints.markAsFailed(endpoint, Deadline.now + settings.RetryGateClosedFor)
-        case (Some(Pass(endpoint, Some(currentUid), _)), Some(quarantineUid)) if currentUid == quarantineUid ⇒
-          context.stop(endpoint)
-        case _ ⇒
-        // Do nothing, because either:
-        // A: we don't know yet the UID of the writer, it will be checked against current quarantine state later
-        // B: we know the UID, but it does not match with the UID to be quarantined
+        case (Some(Pass(endpoint, uidOption, refuseUidOption)), Some(quarantineUid)) ⇒
+          uidOption match {
+            case Some(uid) if uid == quarantineUid ⇒
+              endpoints.markAsQuarantined(address, quarantineUid, Deadline.now + settings.QuarantineDuration)
+              eventPublisher.notifyListeners(QuarantinedEvent(address, quarantineUid))
+              context.stop(endpoint)
+            // or it does not match with the UID to be quarantined
+            case None if !refuseUidOption.contains(quarantineUid) ⇒
+              // the quarantine uid may be got fresh by cluster gossip, so update refuseUid for late handle when the writer got uid
+              endpoints.registerWritableEndpointRefuseUid(address, quarantineUid)
+            case _ ⇒ //the quarantine uid has lost the race with some failure, do nothing
+          }
+        case (_, Some(quarantinedUid)) ⇒
+          // the current state is gated or quarantined, and we know the UID, update
+          endpoints.markAsQuarantined(address, quarantinedUid, Deadline.now + settings.QuarantineDuration)
+          eventPublisher.notifyListeners(QuarantinedEvent(address, quarantinedUid))
+        case _ ⇒ // the current state is gated or quarantined, and we don't know the UID, do nothing.
       }
 
       // Stop inbound read-only associations
-      (endpoints.readOnlyEndpointFor(address), uidToQuarantineOption) match {
-        case (Some((endpoint, _)), None) ⇒ context.stop(endpoint)
-        case (Some((endpoint, currentUid)), Some(quarantineUid)) if currentUid == quarantineUid ⇒ context.stop(endpoint)
+      endpoints.readOnlyEndpointFor(address) match {
+        case Some((endpoint, currentUid)) if uidToQuarantineOption.forall(_ == currentUid) ⇒ context.stop(endpoint)
```

Comment:
I'd like that we touch as little as possible here, so please revert if this is not solving anything